watch/watch.go

```diff
@@ -0,0 +1,360 @@
+// query changes, a notification will be send back to the user.
+type Watcher struct {
+	id       string
+	store    storage.Store
+	compiler *ast.Compiler
+	ctx      context.Context
+
+	rules []*ast.Rule
+
+	queryWatch map[string]queryNotify         // map from id -> queryNotify
+	dataWatch  map[string]map[signal]struct{} // map from path -> set of signals
+
+	closed bool
+
+	l sync.Mutex // The lock around creating and ending watches.
+}
+
+// Event represents a change to a query. ID is the query ID supplied when the watch on
+// the query was created and Value is the JSON encoded results of the new query
+// evaluation. Error will be populated if evaluating the new query results encountered an
+// error for any reason. If Error is not nil, the contents of Value are undefined.
+type Event struct {
+	ID    string         `json:"id"`
+	Value rego.ResultSet `json:"value"`
+	Error error          `json:"error,omitempty"`
+}
+
+type signal chan struct{}
+
+type queryNotify struct {
+	query string // the original query, used for migration
+
+	out    chan<- Event // channel to send new query results on.
+	notify signal       // channel to recieve new data change alerts on.
+
+	done signal // closed by the watcher to signal the sending goroutine to stop sending query results.
+	ack  signal // closed by the sending goroutine to tell the watcher it has stopped sending query results.
+}
+
+// New creates and returns a new Watcher on the storage.Store s using the ast.Compiler c.
+// Once a compiler is provided to create a Watcher, it must not be modified, or else
+// the results produced by the Watcher are undefined.
+//
+// The Watcher will already be started by New.
+//
+// The ID provided must be unique to the storage.Store.
+func New(ctx context.Context, s storage.Store, c *ast.Compiler, id string, txn storage.Transaction) (*Watcher, error) {
+	w := create(ctx, s, c, id)
+	return w, w.Start(txn)
+}
+
+// Start starts the watcher listening for changes to queries registered on it.
+func (w *Watcher) Start(txn storage.Transaction) error {
+	return w.store.Register(w.ctx, txn, w.id, storage.TriggerConfig{OnCommit: w.notify})
+}
+
+// Query registers a watch on the provided Rego query. Whenever changes are made to a
+// base or virtual document that the query depends on, an Event describing the new result
+// of the query will be sent along the returned channel. The provided id string may be
+// used to end the watch on the query.
+//
+// WatchQuery will return an error if registering the watch fails for any reason.
+func (w *Watcher) Query(query, id string) (<-chan Event, error) {
+	out := make(chan Event)
+	if err := w.query(query, id, out); err != nil {
+		close(out)
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// EndQuery ends the watch on the query associated with id. It will close the channel
+// that was delivering notifications for the given ID. This may happen before or after
+// EndQuery returns.
+func (w *Watcher) EndQuery(id string) error {
```

Comment:
It doesn't have to be asynchronous, but I'm not sure what you mean by check the result? The only way this can fail is if the watch doesn't exist. Channels are already asynchronous by nature, whatever is listening to it should be able to handle it being closed at an arbitrary time. 

That being said, I don't feel really strongly about it, and making this block until the channel is closed is a simple change.